File: typescript/src/model/ShoppingCart.ts
import {Product} from "./Product"
import {SupermarketCatalog} from "./SupermarketCatalog"
import * as _ from "lodash"
import {ProductQuantity} from "./ProductQuantity"
import {Discount} from "./Discount"
import {Receipt} from "./Receipt"
import {Offer} from "./Offer"
import {SpecialOfferType} from "./SpecialOfferType"

type ProductQuantities = { [productName: string]: ProductQuantity }
export type OffersByProduct = {[productName: string]: Offer};

export class ShoppingCart {

    private readonly  items: ProductQuantity[] = [];
    _productQuantities: ProductQuantities = {};


    getItems(): ProductQuantity[] {
        return _.clone(this.items);
    }

    addItem(product: Product): void {
        this.addItemQuantity(product, 1.0);
    }

    productQuantities(): ProductQuantities {
        return this._productQuantities;
    }


    public addItemQuantity(product: Product, quantity: number): void {
        let productQuantity = new ProductQuantity(product, quantity)
        this.items.push(productQuantity);
        let currentQuantity = this._productQuantities[product.name]
        if (currentQuantity) {
            this._productQuantities[product.name] = this.increaseQuantity(product, currentQuantity, quantity);
        } else {
            this._productQuantities[product.name] = productQuantity;
        }

    }

    private increaseQuantity(product: Product, productQuantity: ProductQuantity, quantity: number) {
        return new ProductQuantity(product, productQuantity.quantity + quantity)
    }

    handleOffers(receipt: Receipt,  offers: OffersByProduct, catalog: SupermarketCatalog ):void {
        for (const productName in this.productQuantities()) {
            const productQuantity = this._productQuantities[productName]
            const product = productQuantity.product;
            const quantity: number = this._productQuantities[productName].quantity;
            if (offers[productName]) {
                const offer : Offer = offers[productName];
                const unitPrice: number= catalog.getUnitPrice(product);
                let quantityAsInt = quantity;
                let discount : Discount|null = null;
                let x = 1;
                if (offer.offerType == SpecialOfferType.ThreeForTwo) {
                    x = 3;

                } else if (offer.offerType == SpecialOfferType.TwoForAmount) {
                    x = 2;
                    if (quantityAsInt >= 2) {
                        const total = offer.argument * Math.floor(quantityAsInt / x) + quantityAsInt % 2 * unitPrice;
                        const discountN = unitPrice * quantity - total;
                        discount = new Discount(product, "2 for " + offer.argument, discountN);
                    }

                } if (offer.offerType == SpecialOfferType.FiveForAmount) {
                    x = 5;
                }
                const numberOfXs = Math.floor(quantityAsInt / x);
                if (offer.offerType == SpecialOfferType.ThreeForTwo && quantityAsInt > 2) {
                    const discountAmount = quantity * unitPrice - ((numberOfXs * 2 * unitPrice) + quantityAsInt % 3 * unitPrice);
                    discount = new Discount(product, "3 for 2", discountAmount);
                }
                if (offer.offerType == SpecialOfferType.TenPercentDiscount) {
                    discount = new Discount(product, offer.argument + "% off", quantity * unitPrice * offer.argument / 100.0);
                }
                if (offer.offerType == SpecialOfferType.FiveForAmount && quantityAsInt >= 5) {
                    const discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * unitPrice);
                    discount = new Discount(product, x + " for " + offer.argument, discountTotal);
                }
                if (discount != null)
                    receipt.addDiscount(discount);
            }

        }
    }
}

File: swift/SupermarketReceipt/SupermarketReceipt/ShoppingCart.swift
public class ShoppingCart {

    public var items = [ProductQuantity]()
    public var productQuantities = [Product: Double]()

    func addItem(product: Product) {
        self.addItemQuantity(product: product, quantity: 1.0)
    }

    public func addItemQuantity(product: Product , quantity: Double) {
        items.append(ProductQuantity(product: product, quantity:quantity))
        if productQuantities[product] != nil {
            productQuantities[product] = productQuantities[product]! + quantity
        } else {
            productQuantities[product] = quantity
        }
    }

    func handleOffers(receipt: Receipt, offers: [Product: Offer], catalog: SupermarketCatalog ) {
        for p in productQuantities.keys {
            var quantity = productQuantities[p]
            if offers[p] != nil {
                var offer = offers[p]
                var unitPrice = catalog.getUnitPrice(product: p)
                var quantityAsInt = Int(quantity!)
                var discount: Discount? = nil
                var x = 1
                if offer?.offerType == SpecialOfferType.ThreeForTwo {
                    x = 3

                } else if offer?.offerType == SpecialOfferType.TwoForAmount {
                    x = 2
                    if (quantityAsInt >= 2) {
                        
                        var intDivision = quantityAsInt / x
                        
                        var pricePerUnit = (offer!.argument * Double(intDivision))
                        
                        var theTotal = Double(quantityAsInt % 2) * unitPrice
                        
                        var total = pricePerUnit + theTotal
                        var discountN = unitPrice * quantity! - total
                        discount =  Discount(description:  "2 for \(offer!.argument)", discountAmount: discountN, product: p)
                    }

                } else if offer?.offerType == SpecialOfferType.FiveForAmount {
                    x = 5
                }
                var numberOfXs = quantityAsInt / x
                if offer?.offerType == SpecialOfferType.ThreeForTwo && quantityAsInt > 2 {
                    var left = Double(numberOfXs * 2) * unitPrice
                    var right = Double(quantityAsInt % 3) * unitPrice
                    var lastPart = left + right
                    var discountAmount = ((quantity ?? 1) * unitPrice) - lastPart
                    discount =  Discount(description: "3 for 2", discountAmount: discountAmount, product: p)
                }
                if offer?.offerType == SpecialOfferType.TenPercentDiscount {
                    discount =  Discount(description: "\(offer!.argument)% off", discountAmount: (quantity ?? 1) * unitPrice * (offer?.argument ?? 1) / 100.0, product: p)
                }
                if offer?.offerType == SpecialOfferType.FiveForAmount && quantityAsInt >= 5 {
                    var left = (unitPrice * (quantity ?? 1))
                    var right = ((offer?.argument ?? 1) * Double(numberOfXs)) + (Double(quantityAsInt % 5) * unitPrice)
                    var discountTotal = left - right
                    discount =  Discount(description: "\(x) for \(offer!.argument)", discountAmount: discountTotal,  product: p)
                }
                if discount != nil {
                    receipt.addDiscount(discount: discount!)
                }
            }

        }
    }
}
